In [3]:
import pandas as pd
import numpy as np
#import matplotlib.pyplot as plt
#import seaborn as sns

In [4]:
df = pd.read_csv('total_data(2016제외) (1).csv')

### xdata로 모든 column을 넣어서 Lasso
- Lasso regression feature selection 효과가 존재하므로 적절한 점수를 가지는 모델에서 weight가 0이 아닌 feature를 선택

In [5]:
# 필요 패키지 로드
#from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Lasso
from sklearn.metrics import mean_absolute_error

In [6]:
# y값인 q1-q5가 결측인 2020년 데이터 제거
a = df[0:-82]
a

jur_stn       year  report_sx_남성  report_sx_불상  report_sx_여성  \
0     서울중부경찰서  2017(상반기)       13404.0        3239.0        5931.0   
1     서울종로경찰서  2017(상반기)       11579.0        2560.0        5534.0   
2    서울남대문경찰서  2017(상반기)        9093.0        1855.0        3551.0   
3    서울서대문경찰서  2017(상반기)       18441.0       10029.0       10298.0   
4     서울혜화경찰서  2017(상반기)       10829.0        2989.0        5038.0   
..        ...        ...           ...           ...           ...   
241   창원서부경찰서  2019(하반기)       12630.0        2990.0        7151.0   
242   마산중부경찰서  2019(하반기)       12223.0        2552.0        6489.0   
243   마산동부경찰서  2019(하반기)       15177.0        3190.0        7647.0   
244     진주경찰서  2019(하반기)       21840.0        5262.0       11455.0   
245     진해경찰서  2019(하반기)       11600.0        2442.0        6278.0   

     report_sx_total  vio_cnt  mur_rob_cnt  car_cnt  5m_crm_yn  ...  \
0            22574.0   1324.0          0.0   1412.0     3275.0  ...   
1            19674.0   1102.0          0.0    867.0     2914.0  ...   
2            14499.0    949.0          0.0    736.0     2506.0  ...   
3            38769.0   1924.0          0.0   1788.0     4267.0  ...   
4            18857.0   1127.0          0.0    899.0     2864.0  ...   
..               ...      ...          ...      ...        ...  ...   
241          22771.0   1460.0          2.0   1964.0     2142.0  ...   
242          21264.0   1265.0          5.0   1584.0     1369.0  ...   
243          26014.0   1715.0          8.0   1910.0     1920.0  ...   
244          38557.0   2552.0          0.0   3570.0     2876.0  ...   
245          20320.0   1444.0          0.0   1692.0     1354.0  ...   

     single_f_2030  single_f_4050  single_f_o60  adult  bell    q1    q2  \
0           2128.0         1230.5        1635.5    0.0   0.0  75.2  76.5   
1           2703.0         1378.5        1782.5    2.0   2.0  81.6  82.4   
2           2128.0         1230.5        1635.5    0.0   0.0  77.3  79.3   
3          12420.0         4496.0        7066.0    0.0   0.0  78.0  80.0   
4           2703.0         1378.5        1782.5    0.0   0.0  78.6  80.6   
..             ...            ...           ...    ...   ...   ...   ...   
241         2900.0         4492.0        5574.0    1.0   1.0  79.1  79.5   
242         2216.0         2837.0        7186.0    3.0   3.0  80.8  80.0   
243         1496.0         2939.0        6166.0    1.0   1.0  75.0  76.5   
244         6026.0         5205.0       11823.0    8.0   8.0  78.7  79.6   
245         1782.0         2592.0        4931.0    3.0   3.0  76.6  79.2   

       q3    q4    q5  
0    69.9  65.9  72.6  
1    69.1  62.7  72.2  
2    68.8  61.0  71.7  
3    67.0  66.3  71.4  
4    70.7  65.2  75.4  
..    ...   ...   ...  
241  68.5  71.2  76.0  
242  70.8  73.2  76.5  
243  69.4  68.9  73.1  
244  70.3  74.6  75.6  
245  67.0  70.9  73.5  

[246 rows x 75 columns]

In [7]:
# 경찰서와 연도 데이터 제거
a.drop(columns = ['jur_stn', 'year'], inplace = True)

C:\Anaconda3\lib\site-packages\pandas\core\frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [8]:
a_ = a.drop(columns = ['q1', 'q2', 'q3', 'q4', 'q5'])

In [9]:
# StandardScaling
scaler = StandardScaler()
scaler.fit(a_) # scaler에 xdata 학습
a_s = scaler.transform(a_)
a_s

array([[-0.87369027, -0.8125045 , -1.08891897, ..., -1.55004946,
        -0.59342342, -0.59342342],
       [-1.07373936, -1.09735738, -1.17135971, ..., -1.50210803,
         0.29943384,  0.29943384],
       [-1.34624459, -1.39311773, -1.58314811, ..., -1.55004946,
        -0.59342342, -0.59342342],
       ...,
       [-0.67934122, -0.83306089, -0.7325756 , ..., -0.07250741,
        -0.14699479, -0.14699479],
       [ 0.05102978,  0.03618089,  0.05819103, ...,  1.77242259,
         2.97800563,  2.97800563],
       [-1.07143743, -1.14686053, -1.01686119, ..., -0.47528071,
         0.74586247,  0.74586247]])

In [10]:
# 2017, 2018 데이터를 학습용, 2019년 데이터를 검증용 데이터셋으로 설정
train_ = a_s[:-82]
test_ = a_s[-82:]
train = a[:-82]
test = a[-82:]

In [11]:
xtrain = train_
ytrain1 = train['q1']
ytrain2 = train['q2']
ytrain3 = train['q3']
ytrain4 = train['q4']
ytrain5 = train['q5']

In [12]:
xtest = test_
ytest1 = test['q1']
ytest2 = test['q2']
ytest3 = test['q3']
ytest4 = test['q4']
ytest5 = test['q5']

In [13]:
a_.columns

Index(['report_sx_남성', 'report_sx_불상', 'report_sx_여성', 'report_sx_total',
       'vio_cnt', 'mur_rob_cnt', 'car_cnt', '5m_crm_yn', 'theft_violence_x',
       'robber_murder_x', 'traffic_x', 'public_x', 'crm', 'theft_violence_y',
       'robber_murder_y', 'traffic_y', 'public_y', 'light_cnt', '차대사람', '차대차',
       '차량단독', '가해_성별_기타불명', '가해_남', '가해_여', '가해_10대이하', '가해_20대', '가해_30대',
       '가해_40대', '가해_50대', '가해_60대', '가해_70대', '가해_80대이상', '가해_미분류', '피해_남',
       '피해_미분류', '피해_여', '피해_10대이하', '피해_20대', '피해_30대', '피해_40대', '피해_50대',
       '피해_60대', '피해_70대', '피해_80대이상', 'cctv', 'pol_num', 'popu_합계', 'for_합계',
       'popu_u20', 'popu_2030', 'popu_4050', 'popu_o60', 'for_u20', 'for_2030',
       'for_4050', 'for_o60', 'single_합계', 'single_u20', 'single_2030',
       'single_4050', 'single_o60', 'single_f_합계', 'single_f_u20',
       'single_f_2030', 'single_f_4050', 'single_f_o60', 'adult', 'bell'],
      dtype='object')

In [14]:
# 그리드 서치 패키지
from sklearn.model_selection import GridSearchCV

In [15]:
# 최적 성능을 내는 lasso의 alpha값을 얻기 위해 param_grid 생성

num = []
for i in range(1, 2000):
    num.append(i * 0.001)

param_grid = {'alpha' : num}

#그리드 서치 설정
grid_search = GridSearchCV(Lasso(), param_grid = param_grid, cv = 10, n_jobs = -1, scoring ='r2')
#grid_search = GridSearchCV(Lasso(), param_grid = param_grid, cv = 10, n_jobs = -1, scoring ='neg_mean_absolute_error')

#### q1 절도폭력 

In [16]:
# 그리드 서치 후 최고 성능의 모델을 lasso1에 저장

grid_search.fit(xtrain, ytrain1)
lasso1 = grid_search.best_estimator_

In [17]:
# MAE 출력

y_pred1 = lasso1.predict(xtest)
mean_absolute_error(ytest1, y_pred1)

2.534933292697339

In [18]:
# 결과

print('alpa =', lasso1.alpha)
# print(lasso1.coef_) # Lasso 회귀분석으로 나온 weghit값

print(a_.columns[lasso1.coef_ != 0]) # Lasso 회귀분석 결과 weight가 0이 아닌 xfeature 리스트
print('feature 갯수:', len(a_.columns[lasso1.coef_ != 0]))
print('가장 강한 양의 상관관계: ',a_.columns[lasso1.coef_.argmax()], '\n가장 강한 음의 상관관계: ', a_.columns[lasso1.coef_.argmin()])

alpa = 0.061
Index(['report_sx_불상', 'vio_cnt', 'mur_rob_cnt', 'robber_murder_x', 'public_y',
       'light_cnt', '가해_성별_기타불명', '가해_10대이하', '가해_20대', '가해_70대', '피해_미분류',
       '피해_여', '피해_10대이하', '피해_30대', '피해_70대', '피해_80대이상', 'pol_num',
       'popu_u20', 'for_2030', 'for_4050', 'single_u20', 'single_4050',
       'single_f_2030', 'adult', 'bell'],
      dtype='object')
feature 갯수: 25
가장 강한 양의 상관관계:  popu_u20 
가장 강한 음의 상관관계:  single_4050


In [19]:
# q1 관련 feature
a_.columns[lasso1.coef_ != 0]

Index(['report_sx_불상', 'vio_cnt', 'mur_rob_cnt', 'robber_murder_x', 'public_y',
       'light_cnt', '가해_성별_기타불명', '가해_10대이하', '가해_20대', '가해_70대', '피해_미분류',
       '피해_여', '피해_10대이하', '피해_30대', '피해_70대', '피해_80대이상', 'pol_num',
       'popu_u20', 'for_2030', 'for_4050', 'single_u20', 'single_4050',
       'single_f_2030', 'adult', 'bell'],
      dtype='object')

#### q2 강도살인

In [20]:
# 그리드 서치 후 최고 성능의 모델을 lasso2에 저장

grid_search.fit(xtrain, ytrain2)
lasso2 = grid_search.best_estimator_

In [21]:
# MAE 출력

y_pred2 = lasso2.predict(xtest)
mean_absolute_error(ytest2, y_pred2)

2.403205855220829

In [22]:
# 결과

print('alpa =', lasso2.alpha)
# print(lasso1.coef_) # Lasso 회귀분석으로 나온 weghit값

print(a_.columns[lasso2.coef_ != 0]) # Lasso 회귀분석 결과 weight가 0이 아닌 xfeature 리스트
print('feature 갯수:', len(a_.columns[lasso2.coef_ != 0]))
print('가장 강한 양의 상관관계: ',a_.columns[lasso2.coef_.argmax()], '\n가장 강한 음의 상관관계: ', a_.columns[lasso2.coef_.argmin()])

alpa = 0.026000000000000002
Index(['vio_cnt', 'car_cnt', 'robber_murder_x', 'traffic_x', 'public_x', 'crm',
       'traffic_y', 'public_y', 'light_cnt', '차대사람', '가해_성별_기타불명', '가해_여',
       '가해_10대이하', '가해_20대', '가해_40대', '가해_70대', '가해_80대이상', '피해_30대',
       '피해_40대', '피해_50대', '피해_70대', '피해_80대이상', 'popu_u20', 'popu_o60',
       'for_u20', 'for_2030', 'for_o60', 'single_u20', 'single_2030',
       'single_4050', 'single_f_2030', 'adult'],
      dtype='object')
feature 갯수: 32
가장 강한 양의 상관관계:  popu_u20 
가장 강한 음의 상관관계:  vio_cnt


In [23]:
# q2 관련 feature
a_.columns[lasso2.coef_ != 0]

Index(['vio_cnt', 'car_cnt', 'robber_murder_x', 'traffic_x', 'public_x', 'crm',
       'traffic_y', 'public_y', 'light_cnt', '차대사람', '가해_성별_기타불명', '가해_여',
       '가해_10대이하', '가해_20대', '가해_40대', '가해_70대', '가해_80대이상', '피해_30대',
       '피해_40대', '피해_50대', '피해_70대', '피해_80대이상', 'popu_u20', 'popu_o60',
       'for_u20', 'for_2030', 'for_o60', 'single_u20', 'single_2030',
       'single_4050', 'single_f_2030', 'adult'],
      dtype='object')

#### q3 교통안전

In [24]:
# 그리드 서치 후 최고 성능의 모델을 lasso3에 저장

grid_search.fit(xtrain, ytrain3)
lasso3 = grid_search.best_estimator_

In [25]:
# MAE 출력

y_pred3 = lasso3.predict(xtest)
mean_absolute_error(ytest3, y_pred3)

2.3832876315760934

In [26]:
# 결과

print('alpa =', lasso3.alpha)
# print(lasso1.coef_) # Lasso 회귀분석으로 나온 weghit값

print(a_.columns[lasso3.coef_ != 0]) # Lasso 회귀분석 결과 weight가 0이 아닌 xfeature 리스트
print('feature 갯수:', len(a_.columns[lasso3.coef_ != 0]))
print('가장 강한 양의 상관관계: ',a_.columns[lasso3.coef_.argmax()], '\n가장 강한 음의 상관관계: ', a_.columns[lasso3.coef_.argmin()])

alpa = 0.132
Index(['traffic_x', 'public_y', 'light_cnt', '가해_10대이하', '가해_70대', '가해_80대이상',
       '피해_10대이하', '피해_70대', 'cctv', 'pol_num', 'for_2030', 'single_4050'],
      dtype='object')
feature 갯수: 12
가장 강한 양의 상관관계:  피해_70대 
가장 강한 음의 상관관계:  single_4050


In [33]:
# q3 관련 feature
a_.columns[lasso3.coef_ != 0]

Index(['traffic_x', 'public_y', 'light_cnt', '가해_10대이하', '가해_70대', '가해_80대이상',
       '피해_10대이하', '피해_70대', 'cctv', 'pol_num', 'for_2030', 'single_4050'],
      dtype='object')

#### q4 법질서 준수도

In [27]:
# 그리드 서치 후 최고 성능의 모델을 lasso4에 저장

grid_search.fit(xtrain, ytrain4)
lasso4 = grid_search.best_estimator_

In [28]:
# MAE 출력

y_pred4 = lasso4.predict(xtest)
mean_absolute_error(ytest4, y_pred4)

3.12443634055981

In [29]:
# 결과

print('alpa =', lasso4.alpha)
# print(lasso1.coef_) # Lasso 회귀분석으로 나온 weghit값

print(a_.columns[lasso4.coef_ != 0]) # Lasso 회귀분석 결과 weight가 0이 아닌 xfeature 리스트
print('feature 갯수:', len(a_.columns[lasso4.coef_ != 0]))
print('가장 강한 양의 상관관계: ',a_.columns[lasso4.coef_.argmax()], '\n가장 강한 음의 상관관계: ', a_.columns[lasso4.coef_.argmin()])

alpa = 0.031
Index(['report_sx_불상', 'vio_cnt', 'mur_rob_cnt', 'car_cnt', '5m_crm_yn',
       'theft_violence_x', 'robber_murder_x', 'theft_violence_y',
       'robber_murder_y', 'traffic_y', 'public_y', 'light_cnt', '가해_성별_기타불명',
       '가해_10대이하', '가해_20대', '가해_40대', '가해_70대', '가해_80대이상', '피해_미분류',
       '피해_10대이하', '피해_50대', '피해_70대', '피해_80대이상', 'cctv', 'pol_num',
       'popu_u20', 'popu_o60', 'for_u20', 'for_2030', 'for_4050', 'single_u20',
       'single_4050', 'single_f_2030', 'adult', 'bell'],
      dtype='object')
feature 갯수: 35
가장 강한 양의 상관관계:  피해_70대 
가장 강한 음의 상관관계:  pol_num


In [34]:
# q4 관련 feature
a_.columns[lasso4.coef_ != 0]

Index(['report_sx_불상', 'vio_cnt', 'mur_rob_cnt', 'car_cnt', '5m_crm_yn',
       'theft_violence_x', 'robber_murder_x', 'theft_violence_y',
       'robber_murder_y', 'traffic_y', 'public_y', 'light_cnt', '가해_성별_기타불명',
       '가해_10대이하', '가해_20대', '가해_40대', '가해_70대', '가해_80대이상', '피해_미분류',
       '피해_10대이하', '피해_50대', '피해_70대', '피해_80대이상', 'cctv', 'pol_num',
       'popu_u20', 'popu_o60', 'for_u20', 'for_2030', 'for_4050', 'single_u20',
       'single_4050', 'single_f_2030', 'adult', 'bell'],
      dtype='object')

#### q5 전반적 안전도

In [30]:
# 그리드 서치 후 최고 성능의 모델을 lasso4에 저장

grid_search.fit(xtrain, ytrain5)
lasso5 = grid_search.best_estimator_

In [31]:
# MAE 출력

y_pred5 = lasso5.predict(xtest)
mean_absolute_error(ytest5, y_pred5)

2.349911049054337

In [32]:
# 결과

print('alpa =', lasso5.alpha)
# print(lasso1.coef_) # Lasso 회귀분석으로 나온 weghit값

print(a_.columns[lasso5.coef_ != 0]) # Lasso 회귀분석 결과 weight가 0이 아닌 xfeature 리스트
print('feature 갯수:', len(a_.columns[lasso5.coef_ != 0]))
print('가장 강한 양의 상관관계: ',a_.columns[lasso5.coef_.argmax()], '\n가장 강한 음의 상관관계: ', a_.columns[lasso5.coef_.argmin()])

alpa = 0.025
Index(['report_sx_불상', 'vio_cnt', 'mur_rob_cnt', 'car_cnt', 'robber_murder_x',
       'crm', 'theft_violence_y', 'traffic_y', 'public_y', 'light_cnt', '차대사람',
       '가해_성별_기타불명', '가해_10대이하', '가해_20대', '가해_40대', '가해_70대', '피해_미분류',
       '피해_10대이하', '피해_70대', '피해_80대이상', 'cctv', 'popu_u20', 'for_u20',
       'for_2030', 'for_4050', 'for_o60', 'single_4050', 'single_o60',
       'single_f_2030', 'adult', 'bell'],
      dtype='object')
feature 갯수: 31
가장 강한 양의 상관관계:  popu_u20 
가장 강한 음의 상관관계:  single_4050


In [35]:
# q5 관련 feature
a_.columns[lasso5.coef_ != 0]

Index(['report_sx_불상', 'vio_cnt', 'mur_rob_cnt', 'car_cnt', 'robber_murder_x',
       'crm', 'theft_violence_y', 'traffic_y', 'public_y', 'light_cnt', '차대사람',
       '가해_성별_기타불명', '가해_10대이하', '가해_20대', '가해_40대', '가해_70대', '피해_미분류',
       '피해_10대이하', '피해_70대', '피해_80대이상', 'cctv', 'popu_u20', 'for_u20',
       'for_2030', 'for_4050', 'for_o60', 'single_4050', 'single_o60',
       'single_f_2030', 'adult', 'bell'],
      dtype='object')

### Lasso를 통한 Feature Selection 결과

- q1  
alpa = 0.061  
        ['report_sx_불상', 'vio_cnt', 'mur_rob_cnt', 'robber_murder_x', 'public_y',
       'light_cnt', '가해_성별_기타불명', '가해_10대이하', '가해_20대', '가해_70대', '피해_미분류',
       '피해_여', '피해_10대이하', '피해_30대', '피해_70대', '피해_80대이상', 'pol_num',
       'popu_u20', 'for_2030', 'for_4050', 'single_u20', 'single_4050',
       'single_f_2030', 'adult', 'bell']
feature 갯수: 25  
가장 강한 양의 상관관계:  popu_u20  
가장 강한 음의 상관관계:  single_4050  

- q2  
alpa = 0.026
        ['vio_cnt', 'car_cnt', 'robber_murder_x', 'traffic_x', 'public_x', 'crm',
       'traffic_y', 'public_y', 'light_cnt', '차대사람', '가해_성별_기타불명', '가해_여',
       '가해_10대이하', '가해_20대', '가해_40대', '가해_70대', '가해_80대이상', '피해_30대',
       '피해_40대', '피해_50대', '피해_70대', '피해_80대이상', 'popu_u20', 'popu_o60',
       'for_u20', 'for_2030', 'for_o60', 'single_u20', 'single_2030',
       'single_4050', 'single_f_2030', 'adult']  
feature 갯수: 32  
가장 강한 양의 상관관계:  popu_u20  
가장 강한 음의 상관관계:  vio_cnt

- q3  
alpa = 0.132  
        ['traffic_x', 'public_y', 'light_cnt', '가해_10대이하', '가해_70대', '가해_80대이상',
       '피해_10대이하', '피해_70대', 'cctv', 'pol_num', 'for_2030', 'single_4050']  
feature 갯수: 12  
가장 강한 양의 상관관계:  피해_70대  
가장 강한 음의 상관관계:  single_4050

- q4  
alpa = 0.031  
        ['report_sx_불상', 'vio_cnt', 'mur_rob_cnt', 'car_cnt', '5m_crm_yn',
       'theft_violence_x', 'robber_murder_x', 'theft_violence_y',
       'robber_murder_y', 'traffic_y', 'public_y', 'light_cnt', '가해_성별_기타불명',
       '가해_10대이하', '가해_20대', '가해_40대', '가해_70대', '가해_80대이상', '피해_미분류',
       '피해_10대이하', '피해_50대', '피해_70대', '피해_80대이상', 'cctv', 'pol_num',
       'popu_u20', 'popu_o60', 'for_u20', 'for_2030', 'for_4050', 'single_u20',
       'single_4050', 'single_f_2030', 'adult', 'bell']  
feature 갯수: 35  
가장 강한 양의 상관관계:  피해_70대  
가장 강한 음의 상관관계:  pol_num

- q5  
alpa = 0.025  
        ['report_sx_불상', 'vio_cnt', 'mur_rob_cnt', 'car_cnt', 'robber_murder_x',
       'crm', 'theft_violence_y', 'traffic_y', 'public_y', 'light_cnt', '차대사람',
       '가해_성별_기타불명', '가해_10대이하', '가해_20대', '가해_40대', '가해_70대', '피해_미분류',
       '피해_10대이하', '피해_70대', '피해_80대이상', 'cctv', 'popu_u20', 'for_u20',
       'for_2030', 'for_4050', 'for_o60', 'single_4050', 'single_o60',
       'single_f_2030', 'adult', 'bell']  
feature 갯수: 31  
가장 강한 양의 상관관계:  popu_u20  
가장 강한 음의 상관관계:  single_4050